discojs/discojs-core/src/models/gpt/index.ts

```diff
@@ -0,0 +1,136 @@
+import tf from '@tensorflow/tfjs'
+
+import { WeightsContainer } from '../..'
+import type { Dataset } from '../../dataset'
+import { Sink } from '../../utils/event_emitter'
+
+import type { EpochLogs, Prediction, Sample } from '../model'
+import { Model } from '../model'
+
+import { GPTLMHeadModel } from './model'
+
+// TODO too big config
+interface Config {
+  modelType: string
+  epochs: number // TODO mv to Task
+  maxIter: number
+  blockSize: number
+  lr: number
+  vocabSize: number
+  maxEvalBatches: number
+}
+
+export class GPT extends Model {
+  private readonly model: GPTLMHeadModel
+
+  private static readonly blockSize = 128
+  private static readonly vocabSize = 50258
+
+  constructor () {
+    super()
+
+    // TODO sensible defaults?
+    const config: Config = {
+      modelType: 'gpt-nano',
+      epochs: 1,
+      maxIter: 2,
+      blockSize: GPT.blockSize,
+      lr: 0.001,
+      vocabSize: GPT.vocabSize,
+      maxEvalBatches: 1
+    }
+
+    this.model = new GPTLMHeadModel(config)
+  }
+
+  override get weights (): WeightsContainer {
+    return new WeightsContainer(this.model.weights.map((w) => w.read()))
+  }
+
+  override set weights (ws: WeightsContainer) {
+    this.model.setWeights(ws.weights)
+  }
+
+  private convertCharDataset (dataset: Dataset): Dataset {
+    const batchSize = 4
+    const sampleSize = GPT.blockSize + 1
+    const chunkSize = sampleSize * batchSize * 2
+
+    function toUInt16 (low: number, high: number): number {
+      low &= 0xff
+      high &= 0xff
+      return (high << 8) | low
+    }
+
+    // TODO add support for small last batch
+    return dataset.batch(chunkSize, false).mapAsync(async (chunk) => {
+      if (!(chunk instanceof tf.Tensor)) {
+        throw new Error('chunk is not a Tensor')
+      }
+      if (chunk.shape.length !== 2 || chunk.shape[1] !== 1) {
+        throw new Error('dataset is not a only char')
+      }
+
+      const buffer = await chunk.buffer()
+
+      const xs = tf.buffer([batchSize, GPT.blockSize], 'int32')
+      const ys = tf.buffer([batchSize, GPT.blockSize, GPT.vocabSize], 'int32')
+
+      for (let i = 0; i < batchSize; i++) {
+        for (let j = 0; j < sampleSize; j++) {
+          const idx = (i * sampleSize + j) * 2
+          const low = buffer.get(idx)
+          const high = buffer.get(idx + 1)
+          const token = toUInt16(low, high)
+          if (j < sampleSize - 1) xs.set(token, i, j)
+          if (j > 0) ys.set(1, i, j - 1, token)
+        }
+      }
+
+      return { xs: xs.toTensor(), ys: ys.toTensor() }
+    })
+  }
+
+  override async * train (
+    trainingData: Dataset,
+    validationData?: Dataset,
+    epochs = 1,
+    tracker = new Sink()
+  ): AsyncGenerator<EpochLogs, void> {
+    for (let i = 0; i < epochs; i++) {
+      let logs: tf.Logs | undefined
+
+      await this.model.fitDataset(
+        this.convertCharDataset(trainingData), {
+          epochs: 1,
```

Comment:
Also, do you know why we need to manually iterate over the epochs rather than using the `epochs` parameter?
When I remove the for-loop and use the `epochs` parameter, the training stops after the first epoch
(@peacefulotter)

Reply:
I am not sure why training stops after one epoch. If I remember correctly, I had to bypass the batchSize parameter because the text loader creates the batch itself (without using gf dataset.batch(batchSize). This was done for improving performance. However, I am not sure it has a link with epochs..

Reply:
whichever is easiest...
in LLM terminology, one datapoint (=element of the batch) would usually be the whole sequence (of fixed context length, say 500, with padding the non-used tokens). batch-sizes larger than 1 would be nice to support, but it has to be its own tensor dimension independent of the sequence length

Reply:
I am not sure I made myself clear but batches are implemented and work. The only thing is that the text loader doesn't use the traditional .batch(batchSize) but rather constructs it itself

Reply:
`Model` is using generators instead of callbacks, to be able to stop it from the outside and to get logs from every epoch. as such, each `fitDataset` call is configured to do a single epoch, but as the weights are not reset, `Model` is effectively running multiple epoch.

there are issue when using the whole dataset (some out of bound access), so the current code is only using the first `config.maxIter` batches.

Reply:
Ah thanks! I'm surprised this is necessary to be able to stop and get training statistics though.